File: clients/cpp/tests/conformance.cpp
#include <cstdint>
#include <exception>
#include <iostream>
#include <span>
#include <stdexcept>
#include <string>
#include <string_view>
#include <thread>
#include <vector>

#include <darkinno/crdt_rga.hpp>

namespace {

using darkinno::crdt::Error;
using darkinno::crdt::LwwMap;
using darkinno::crdt::Rga;

void Require(bool condition, std::string_view message) {
  if (!condition) {
    throw std::runtime_error(std::string(message));
  }
}

std::vector<std::uint8_t> Hex(std::string_view value) {
  Require(value.size() % 2 == 0, "hex input must have complete bytes");
  std::vector<std::uint8_t> output;
  output.reserve(value.size() / 2);
  for (std::size_t index = 0; index < value.size(); index += 2) {
    const auto high = value[index];
    const auto low = value[index + 1];
    const auto nibble = [](char character) -> std::uint8_t {
      if (character >= '0' && character <= '9') {
        return static_cast<std::uint8_t>(character - '0');
      }
      if (character >= 'a' && character <= 'f') {
        return static_cast<std::uint8_t>(character - 'a' + 10);
      }
      throw std::runtime_error("invalid hexadecimal digit");
    };
    output.push_back(static_cast<std::uint8_t>((nibble(high) << 4U) | nibble(low)));
  }
  return output;
}

std::span<const std::uint8_t> Bytes(const std::vector<std::uint8_t>& value) {
  return std::span<const std::uint8_t>(value.data(), value.size());
}

}  // namespace

int main() {
  try {
    const auto vector = Hex("435244540114001201010205616c696365000100410101b20700c1d69811");
    Rga vector_reader("cpp-vector-reader");
    vector_reader.Apply(Bytes(vector));
    Require(vector_reader.Text() == "Aβ", "Go canonical vector projected the wrong text");

    const auto before = vector_reader.Text();
    auto corrupt = vector;
    corrupt.back() ^= 0x01U;
    try {
      vector_reader.Apply(Bytes(corrupt));
      throw std::runtime_error("corrupt frame was accepted");
    } catch (const Error& error) {
      Require(error.Code() == CRDT_INVALID_FRAME, "corrupt frame returned the wrong status");
    }
    Require(vector_reader.Text() == before, "corrupt frame changed visible text");

    Rga alice("cpp-alice");
    Rga bob("cpp-bob");
    Rga carol("cpp-carol");
    const auto initial = alice.Insert(0, "A");
    bob.Apply(Bytes(initial));
    carol.Apply(Bytes(initial));
    const auto bob_edit = bob.Insert(1, "B");
    const auto carol_edit = carol.Insert(1, "C");
    for (const auto* frame : {&carol_edit, &bob_edit, &bob_edit}) {
      alice.Apply(Bytes(*frame));
    }
    bob.Apply(Bytes(carol_edit));
    carol.Apply(Bytes(bob_edit));
    Require(alice.Text() == bob.Text() && alice.Text() == carol.Text(),
            "duplicate/reordered replicas did not converge");

    Rga recovered(alice.Clock());
    recovered.Apply(Bytes(alice.State()));
    Require(recovered.Text() == alice.Text(), "state recovery did not converge");
    const auto recovered_edit = recovered.Insert(recovered.Text().size(), "D");
    alice.Apply(Bytes(recovered_edit));
    Require(recovered.Text() == alice.Text(),
            "same-ID clock recovery produced a conflicting mutation");

    const crdt_limits rga_limits{
        .max_frame_bytes = 1U << 20U,
        .max_payload_bytes = 1U << 20U,
        .max_string_bytes = 64U << 10U,
        .max_nodes = 1,
        .max_tags = 1,
        .max_tombstones = 1,
        .max_pending_nodes = 1,
        .max_pending_bytes = 1,
    };
    Rga limited_rga("cpp-limited", rga_limits);
    [[maybe_unused]] const auto limited_initial = limited_rga.Insert(0, "A");
    const auto limited_state = limited_rga.State();
    Rga limited_recovered(limited_rga.Clock(), rga_limits);
    limited_recovered.Apply(Bytes(limited_state));
    Rga unbounded_rga("cpp-unbounded");
    const auto excess_delta = unbounded_rga.Insert(0, "B");
    const auto rga_limit_state = limited_recovered.State();
    const auto rga_limit_clock = limited_recovered.Clock();
    try {
      limited_recovered.Apply(Bytes(excess_delta));
      throw std::runtime_error("recovered RGA accepted a frame above its node limit");
    } catch (const Error& error) {
      Require(error.Code() == CRDT_RESOURCE_LIMIT, "RGA limit returned the wrong status");
    }
    Require(limited_recovered.State() == rga_limit_state,
            "RGA over-limit frame changed recovered state");
    const auto actual_rga_limit_clock = limited_recovered.Clock();
    Require(actual_rga_limit_clock.replica_id == rga_limit_clock.replica_id &&
                actual_rga_limit_clock.wall_time == rga_limit_clock.wall_time &&
                actual_rga_limit_clock.logical == rga_limit_clock.logical,
            "RGA over-limit frame changed recovered HLC");

    const auto map_vector = Hex("435244540109000e01016105616c69636501000101783c3edf37");
    LwwMap map_reader("cpp-map-vector-reader");
    map_reader.Apply(Bytes(map_vector));
    const auto map_value = map_reader.Get("a");
    Require(map_value.has_value() && *map_value == std::vector<std::uint8_t>{'x'},
            "Go LWW-Map vector projected the wrong value");
    Require(map_reader.State() == map_vector, "Go LWW-Map state did not re-encode canonically");

    LwwMap map_alice("cpp-map-alice");
    LwwMap map_bob("cpp-map-bob");
    LwwMap map_carol("cpp-map-carol");
    const auto map_initial = map_alice.Set("title", Bytes(std::vector<std::uint8_t>{'d', 'r', 'a', 'f', 't'}));
    map_bob.Apply(Bytes(map_initial));
    map_carol.Apply(Bytes(map_initial));
    const auto map_bob_edit = map_bob.Set("owner", Bytes(std::vector<std::uint8_t>{'b', 'o', 'b'}));
    const auto map_carol_edit = map_carol.Set("title", Bytes(std::vector<std::uint8_t>{'r', 'e', 'v', 'i', 'e', 'w', 'e', 'd'}));
    const auto map_delete = map_alice.Delete("obsolete");
    for (const auto* frame : {&map_carol_edit, &map_bob_edit, &map_delete, &map_bob_edit, &map_initial}) {
      map_alice.Apply(Bytes(*frame));
    }
    for (const auto* frame : {&map_delete, &map_carol_edit, &map_initial, &map_delete}) {
      map_bob.Apply(Bytes(*frame));
    }
    for (const auto* frame : {&map_bob_edit, &map_delete, &map_bob_edit, &map_initial}) {
      map_carol.Apply(Bytes(*frame));
    }
    Require(map_alice.Get("title") == std::optional<std::vector<std::uint8_t>>(
                                         std::vector<std::uint8_t>{'r', 'e', 'v', 'i', 'e', 'w', 'e', 'd'}),
            "LWW-Map did not retain the largest tag");
    Require(map_alice.Keys() == std::vector<std::string>{"owner", "title"},
            "LWW-Map keys were not canonical");
    Require(map_alice.State() == map_bob.State() && map_alice.State() == map_carol.State(),
            "LWW-Map duplicate/reordered replicas did not converge");
    LwwMap recovered_map(map_alice.Clock());
    recovered_map.Apply(Bytes(map_alice.State()));
    const auto recovered_map_edit = recovered_map.Set(
        "after-recovery", Bytes(std::vector<std::uint8_t>{'s', 'a', 'f', 'e'}));
    map_alice.Apply(Bytes(recovered_map_edit));
    Require(map_alice.State() == recovered_map.State(),
            "LWW-Map same-ID recovery produced a conflicting mutation");

    LwwMap concurrent_map("cpp-map-concurrent");
    std::vector<std::thread> writers;
    for (std::uint32_t index = 0; index < 4; ++index) {
      writers.emplace_back([&concurrent_map, index] {
        const auto key = "worker:" + std::to_string(index);
        const auto value = std::vector<std::uint8_t>{static_cast<std::uint8_t>('a' + index)};
        [[maybe_unused]] const auto delta = concurrent_map.Set(key, Bytes(value));
      });
    }
    for (auto& writer : writers) {
      writer.join();
    }
    Require(concurrent_map.Keys().size() == 4,
            "mutex-protected LWW-Map handle lost a concurrent local write");

    const crdt_lww_map_limits limited_map_limits{
        .max_frame_bytes = 1U << 20U,
        .max_payload_bytes = 1U << 20U,
        .max_string_bytes = 64U << 10U,
        .max_entries = 1,
        .max_tombstones = 1,
    };
    LwwMap limited_map("cpp-map-limited", limited_map_limits);
    [[maybe_unused]] const auto first_limited_delta =
        limited_map.Set("first", Bytes(std::vector<std::uint8_t>{'s', 'a', 'f', 'e'}));
    const auto limited_before = limited_map.State();
    try {
      [[maybe_unused]] const auto second_limited_delta =
          limited_map.Set("second", Bytes(std::vector<std::uint8_t>{'x'}));
      throw std::runtime_error("over-limit LWW-Map write was accepted");
    } catch (const Error& error) {
      Require(error.Code() == CRDT_RESOURCE_LIMIT, "LWW-Map limit returned the wrong status");
    }
    Require(limited_map.State() == limited_before, "over-limit LWW-Map write changed state");
    std::cout << "PASS: Go vectors, atomic rejection, three-replica convergence, and snapshot recovery for RGA and LWW-Map\n";
  } catch (const std::exception& error) {
    std::cerr << "FAIL: " << error.what() << '\n';
    return 1;
  }
}

File: clients/rust/tests/conformance.rs
use darkinno_crdt_rga::{Delta, Error, Limits, Rga};

const VECTORS: &str = include_str!("../../../docs/protocol/testdata/rga-run-v2-vectors.json");

#[test]
fn go_canonical_vectors_decode_reencode_and_project_identically() {
    for (name, hex, visible) in vector_rows() {
        let frame = hex_bytes(hex);
        let mut document =
            Rga::new(format!("reader-{name}"), Limits::default()).expect("valid replica");
        document
            .apply_frame_at(&frame, 100)
            .expect("Go canonical frame must decode");
        assert_eq!(document.text(), visible, "visible projection for {name}");
        if frame[5] == 20 {
            assert_eq!(
                Delta::decode(&frame, Limits::default())
                    .expect("delta decode")
                    .encode(Limits::default())
                    .expect("delta encode"),
                frame,
                "canonical delta for {name}"
            );
        } else {
            assert_eq!(
                document.encode_state().expect("complete state"),
                frame,
                "canonical state for {name}"
            );
        }
    }
}

#[test]
fn malformed_or_limited_frames_leave_document_unchanged() {
    let (_, valid_hex, _) = vector_rows().into_iter().next().expect("first vector");
    let valid = hex_bytes(valid_hex);
    let mut document = Rga::new("atomic-reader", Limits::default()).expect("valid replica");
    document.apply_frame_at(&valid, 100).expect("valid frame");
    let before_text = document.text();
    let before_state = document.encode_state().expect("complete state");

    let mut corrupt = valid.clone();
    let last = corrupt.len() - 1;
    corrupt[last] ^= 0x01;
    assert_eq!(
        document.apply_frame_at(&corrupt, 101),
        Err(Error::InvalidFrame)
    );
    assert_eq!(document.text(), before_text);
    assert_eq!(
        document.encode_state().expect("unchanged state"),
        before_state
    );

    let limits = Limits {
        max_frame_bytes: valid.len() - 1,
        ..Limits::default()
    };
    let mut bounded = Rga::new("bounded-reader", limits).expect("valid bounds");
    assert_eq!(
        bounded.apply_frame_at(&valid, 101),
        Err(Error::ResourceLimit)
    );
    assert_eq!(bounded.text(), "");
}

#[test]
fn rejected_local_insert_leaves_clock_and_state_unchanged() {
    let limits = Limits {
        max_frame_bytes: 1,
        ..Limits::default()
    };
    let mut document = Rga::new("locally-atomic", limits).expect("valid bounds");
    let before_clock = document.clock_state();

    assert_eq!(
        document.insert_at(0, "A", 10),
        Err(Error::ResourceLimit),
        "the complete frame limit rejects the local edit"
    );
    assert_eq!(document.text(), "");
    assert_eq!(document.clock_state(), before_clock);
}

#[test]
fn duplicate_reordered_and_snapshot_recovery_converge() {
    let limits = Limits::default();
    let mut alice = Rga::new("alice", limits).expect("alice");
    let mut bob = Rga::new("bob", limits).expect("bob");
    let mut carol = Rga::new("carol", limits).expect("carol");

    let initial = alice.insert_at(0, "A", 1).expect("initial edit");
    bob.apply_frame_at(&initial, 1).expect("bob initial");
    carol.apply_frame_at(&initial, 1).expect("carol initial");
    let bob_edit = bob.insert_at(1, "B", 2).expect("bob edit");
    let carol_edit = carol.insert_at(1, "C", 3).expect("carol edit");

    for frame in [&carol_edit, &bob_edit, &bob_edit, &initial] {
        alice.apply_frame_at(frame, 4).expect("alice delivery");
    }
    for frame in [&carol_edit, &initial, &carol_edit] {
        bob.apply_frame_at(frame, 4).expect("bob delivery");
    }
    for frame in [&bob_edit, &bob_edit, &initial] {
        carol.apply_frame_at(frame, 4).expect("carol delivery");
    }

    assert_eq!(alice.text(), "ACB");
    assert_eq!(bob.text(), alice.text());
    assert_eq!(carol.text(), alice.text());

    let snapshot = alice.encode_state().expect("complete snapshot");
    let mut recovered = Rga::new("recovered", limits).expect("recovery replica");
    recovered
        .apply_frame_at(&snapshot, 5)
        .expect("state recovery");
    assert_eq!(recovered.text(), alice.text());
    assert_eq!(
        recovered.encode_state().expect("canonical recovery state"),
        snapshot
    );
}

#[test]
fn child_before_parent_is_bounded_pending_then_integrates() {
    let limits = Limits::default();
    let mut writer = Rga::new("writer", limits).expect("writer");
    let parent = writer.insert_at(0, "A", 1).expect("parent");
    let child = writer.insert_at(1, "β", 2).expect("child");
    let mut receiver = Rga::new("receiver", limits).expect("receiver");

    receiver
        .apply_frame_at(&child, 2)
        .expect("out-of-order child accepted");
    assert_eq!(receiver.pending_count(), 1);
    assert_eq!(receiver.text(), "");
    assert_eq!(receiver.encode_state(), Err(Error::IncompleteState));
    receiver
        .apply_frame_at(&parent, 2)
        .expect("parent resolves child");
    assert_eq!(receiver.pending_count(), 0);
    assert_eq!(receiver.text(), "Aβ");
}

#[test]
fn snapshot_and_clock_recovery_do_not_reuse_local_positions() {
    let limits = Limits::default();
    let mut writer = Rga::new("same-replica", limits).expect("writer");
    writer.insert_at(0, "A", 10).expect("initial edit");
    let state = writer.encode_state().expect("complete state");
    let clock = writer.clock_state();
    let mut recovered = Rga::from_clock_state(clock, limits).expect("restored clock");
    recovered
        .apply_frame_at(&state, 10)
        .expect("restored state");
    let next = recovered
        .insert_at(1, "B", 10)
        .expect("post-recovery local edit");
    writer
        .apply_frame_at(&next, 10)
        .expect("new tag must not conflict");
    assert_eq!(writer.text(), recovered.text());
}

fn vector_rows() -> Vec<(&'static str, &'static str, &'static str)> {
    const NAME: &str = "\"name\": \"";
    const HEX: &str = "\"hex\": \"";
    const VISIBLE: &str = "\"visible_text_after_apply_to_empty\": \"";
    let mut rows = Vec::new();
    let mut remaining = VECTORS;
    while let Some(name_start) = remaining.find(NAME) {
        remaining = &remaining[name_start + NAME.len()..];
        let name = remaining.split_once('"').expect("fixture name").0;
        let hex_start = remaining.find(HEX).expect("fixture hex");
        let after_hex = &remaining[hex_start + HEX.len()..];
        let hex = after_hex.split_once('"').expect("fixture hex value").0;
        let visible_start = after_hex.find(VISIBLE).expect("fixture visible text");
        let after_visible = &after_hex[visible_start + VISIBLE.len()..];
        let visible = after_visible
            .split_once('"')
            .expect("fixture visible value")
            .0;
        rows.push((name, hex, visible));
        remaining = after_visible;
    }
    rows
}

fn hex_bytes(value: &str) -> Vec<u8> {
    assert_eq!(value.len() % 2, 0, "hex has complete bytes");
    value
        .as_bytes()
        .chunks_exact(2)
        .map(|chunk| {
            u8::from_str_radix(std::str::from_utf8(chunk).expect("ASCII hex"), 16)
                .expect("valid fixture hex")
        })
        .collect()
}
